src/message/messageReducers.js

```diff
@@ -0,0 +1,157 @@
+/* @flow */
+import omit from 'lodash.omit';
+
+import type {
+  MessagesState,
+  MessageAction,
+  MessageFetchCompleteAction,
+  EventReactionAddAction,
+  EventReactionRemoveAction,
+  EventNewMessageAction,
+  EventMessageDeleteAction,
+  EventUpdateMessageAction,
+} from '../types';
+import {
+  APP_REFRESH,
+  LOGOUT,
+  LOGIN_SUCCESS,
+  ACCOUNT_SWITCH,
+  MESSAGE_FETCH_COMPLETE,
+  EVENT_NEW_MESSAGE,
+  EVENT_MESSAGE_DELETE,
+  EVENT_REACTION_ADD,
+  EVENT_REACTION_REMOVE,
+  EVENT_UPDATE_MESSAGE,
+} from '../actionConstants';
+import { NULL_ARRAY, NULL_OBJECT } from '../nullObjects';
+import { groupItemsById } from '../utils/misc';
+
+const initialState: MessagesState = NULL_OBJECT;
+
+const messageFetchComplete = (
+  state: MessagesState,
+  action: MessageFetchCompleteAction,
+): MessagesState => ({
+  ...state,
+  ...groupItemsById(action.messages),
+});
+
+const eventReactionAdd = (state: MessagesState, action: EventReactionAddAction): MessagesState => {
+  const oldMessage = state[action.message_id];
+  if (!(action.message_id in state)) {
+    return state;
+  }
+  return {
+    ...state,
+    [action.message_id]: {
+      ...oldMessage,
+      reactions: oldMessage.reactions.concat({
+        emoji_name: action.emoji_name,
+        user: action.user,
+      }),
+    },
+  };
+};
+
+const eventReactionRemove = (
+  state: MessagesState,
+  action: EventReactionRemoveAction,
+): MessagesState => {
+  const oldMessage = state[action.message_id];
+  if (!(action.message_id in state)) {
+    return state;
+  }
+  return {
+    ...state,
+    [action.message_id]: {
+      ...oldMessage,
+      reactions: oldMessage.reactions.filter(
+        x => !(x.emoji_name === action.emoji_name && x.user.email === action.user.email),
+      ),
+    },
+  };
+};
+
+const eventNewMessage = (state: MessagesState, action: EventNewMessageAction): MessagesState => ({
+  ...state,
+  [action.message.id]: action.message,
+});
+
+const eventMessageDelete = (
+  state: MessagesState,
+  action: EventMessageDeleteAction,
+): MessagesState => omit(state, action.messageId);
```

Comment:
`lodash.omit` seems to make a new object even if the key to be omitted is already absent. So this should get the `if (!(action.messageId in state))` treatment too.